[Merged by Bors] - feat(RingTheory): make Euclidean domain from DVR

kckennylau opened the pull request:
---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include at least one commit authored by each
co-author among the commits in the pull request. If necessary, you may 
create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

When merging, all the commits will be squashed into a single commit listing all co-authors.

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

Zulip: https://leanprover.zulipchat.com/#narrow/channel/116395-maths/topic/Leading.20term.20of.20valuation/near/539666529

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

kckennylau committed:
Implement Euclidean domain structure for DVRs

pechersky commented:
In fact, this is stronger -- a DVR is a local PIR. And over local domains, PIR iff Euclidean domain.

kckennylau committed:
Refactor remainder function and related proofs

pechersky commented:
Euclidean domain implies PID. My suggestion is to localize the actual place you use the local ring assumption (and maximal ideal is principal) to the proof, not to the definition of your quotient and remainder. 

Is it the case that the addVal is only defined when you have IsDVR?

kckennylau committed:
doc

Clarify the description of the GCD algorithm in the noncomputable Euclidean domain structure.

kckennylau committed:
docs, reorder

kckennylau committed:
Merge branch 'leanprover-community:master' into dvr-euclidean

pechersky reviewed (approved):
Thanks for adding more information to the docstrings! Consider also adding the reference to `toEuclideanDomain` in the module docstring.

maintainer merge

kckennylau committed:
doc

Added definition for Euclidean domain structure on DVR.

Vierkantor reviewed (approved):
Thanks!

bors r+
